test/Microsoft.AspNetCore.SignalR.Common.Tests/Internal/Protocol/JsonHubProtocolTests.cs

```diff
@@ -10,138 +13,138 @@ namespace Microsoft.AspNetCore.SignalR.Common.Tests.Internal.Protocol
 
-            var protocol = new JsonHubProtocol(new JsonSerializer());
-            var encoded = await protocol.WriteToArrayAsync(message);
-            var json = Encoding.UTF8.GetString(encoded);
+            public double DoubleProp => 6.2831853071;
 
-            Assert.Equal(expectedOutput, json);
-        }
+            public int IntProp => 42;
 
-        [Theory]
-        [InlineData(false, NullValueHandling.Ignore, "{\"invocationId\":\"123\",\"type\":3,\"result\":{\"StringProp\":\"SignalR!\",\"DoubleProp\":6.2831853071,\"IntProp\":42,\"DateTimeProp\":\"2017-04-11T00:00:00\"}}")]
-        [InlineData(true, NullValueHandling.Ignore, "{\"invocationId\":\"123\",\"type\":3,\"result\":{\"stringProp\":\"SignalR!\",\"doubleProp\":6.2831853071,\"intProp\":42,\"dateTimeProp\":\"2017-04-11T00:00:00\"}}")]
-        [InlineData(false, NullValueHandling.Include, "{\"invocationId\":\"123\",\"type\":3,\"result\":{\"StringProp\":\"SignalR!\",\"DoubleProp\":6.2831853071,\"IntProp\":42,\"DateTimeProp\":\"2017-04-11T00:00:00\",\"NullProp\":null}}")]
-        [InlineData(true, NullValueHandling.Include, "{\"invocationId\":\"123\",\"type\":3,\"result\":{\"stringProp\":\"SignalR!\",\"doubleProp\":6.2831853071,\"intProp\":42,\"dateTimeProp\":\"2017-04-11T00:00:00\",\"nullProp\":null}}")]
-        public async Task CanSerializeCustomObjectInCompletion(bool camelCase, NullValueHandling nullHandling, string expectedOutput)
-        {
-            var message = new CompletionMessage("123", error: null, result: new CustomObject(), hasResult: true);
+            public DateTime DateTimeProp => new DateTime(2017, 4, 11);
+
+            public object NullProp => null;
 
-            var jsonSerializer = new JsonSerializer();
-            if(camelCase)
+            public override bool Equals(object obj)
             {
-                jsonSerializer.ContractResolver = new CamelCasePropertyNamesContractResolver();
+                return obj is CustomObject o && Equals(o);
             }
-            jsonSerializer.NullValueHandling = nullHandling;
 
-            var protocol = new JsonHubProtocol(jsonSerializer);
-            var encoded = await protocol.WriteToArrayAsync(message);
-            var json = Encoding.UTF8.GetString(encoded);
-
-            Assert.Equal(expectedOutput, json);
-        }
-
-        private static async Task TestMessageOutput(string expectedOutput, InvocationMessage message, JsonSerializer payloadSerializer = null)
-        {
-            var protocol = new JsonHubProtocol(payloadSerializer ?? new JsonSerializer());
-            var encoded = await protocol.WriteToArrayAsync(message);
-            var json = Encoding.UTF8.GetString(encoded);
+            public override int GetHashCode()
+            {
+                // This is never used in a hash table
+                return 0;
+            }
 
-            Assert.Equal(expectedOutput, json);
+            public bool Equals(CustomObject other)
+            {
```

Comment:
Why do we need it? If we do - is it now a requirement to make custom types equatable if they can be invocation results?